text/0000-feature-metadata.md

```diff
@@ -0,0 +1,309 @@
+add`
+-   In the future, `rustdoc` should not document these features unless
+    `--document-private-items` is specified
+-   A future tool like `cargo info` shouldn't display information about these
+    features
+
+There needs to be an escape hatch for this for things like benchmarks - RFC TBD
+on how this works.
+
+## `unstable`
+
+`unstable` is a boolean value that defaults to `false`. It should indicate that
+a feature gates something that is not API-stable, or can only be built using
+`nightly` Rust.
+
+---
+
+Use cases for this information will likely develop with time, but one of the
+simplest applications is for information output with `cargo add`:
+
+```text
+crab@rust foobar % cargo add regex
+    Updating crates.io index
+      Adding regex v1.7.3 to dependencies.
+             Features:
+             + perf             Enables all performance related features
+             + perf-dfa         Enables the use of a lazy DFA for matching
+             + perf-inline      Enables the use of aggressive inlining inside
+                                match routines
+             + perf-literal     Enables the use of literal optimizations for
+                                speeding up matches
+             + std              When enabled, this will cause regex to use the
+                                standard library
+             + unicode          Enables all Unicode features
+             - unstable (U)     Some unstable features
+             - deprecated (D)   Not a real feature, but it could be
+
+    Updating crates.io index
+```
+
+Features like `aho-corasick`, `memchr`, or `use_std` would likely be `public =
+false` since they aren't listed on the crate landing page.
+
+Another thing to note is sort order. In general, any tool that renders features
+(`rustdoc`, `cargo add`) should attempt to present them in the following way:
+
+-   Display default features first
+-   Display non-default but stable features next (can be in a separate section)
+-   Display unstable or deprecated features last (can be in a separate section)
+-   Do not display private features unless receiving a flag saying to do so (e.g.
+    `--document-private-items` with `rustdoc`)
+-   If ordering is not preserved, present the features alphabetically
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+-   Added complexity to Cargo. Parsing is trivial, but exact implementation
+    details do add test surface area
+-   Added Cargo arguments if escape hatches for `public` or `unstable` are created
+-   Docstrings can be lengthy, adding noise to `Cargo.toml`. This could
+    potentially be solved with the below mentioned `doc-file` key.
+-   `unstable` and `public` uses may not be common enough to be worth including
+-   A markdown parser will eventaully be required to properly parse the `doc`
+    field. Presenting unrendered markdown is likely suitable as a first step,
+    but in general, markdown is required to fully interpret the field.
```

Comment:
A markdown parser is not required to extract the field but to interpret the field and it is required immediately, not eventually.

Presenting unrendered markdown may work in some cases but wouldn't for `cargo add` as we need a markdown parser to extract the summary correctly

Reply:
How about making the field plain text? The `package.description` field doesn't support markdown either, and it simplifies things a lot.

Reply:
Could we say plain text rendering via CLI for now, but markdown rendering in `rustdoc` or tools that support it? This would be consistent with `cargo explain`, which of course has a nice markdown representation in the error code index, but just displays the raw markdown in terminal. Is splitting on `\n\n`/`\r\n\r\n`/EOF not sufficient to extract the summary paragraph?

I'm just worried that defining CLI markdown rendering as part of the MVP may be a bigger blocker than expected. `cargo explain` doesn't render markdown because there is no lightweight tool that does just that - `termimad` [was shut down for being too heavyweight](https://github.com/rust-lang/rust/issues/63128) and I don't know if that same restriction would apply here. (I did start a lightweight parser [here](https://github.com/rust-lang/rust/pull/104540) but haven't yet finished it, and it may be more limited than desired for Cargo output anyway)

Reply:
The questions we need to ask are
- Would rustdoc want to support rendering this with markdown?
  - If not, then that is another drawback compared to solutions like `document-features`
  - If it is, then it should be specified and treated as markdown, rather than rustdoc's rendering leading to people writing markdown in the fields and provide a worse experience for exeryone else
- Should we support an implicit summary, for `cargo add`, tooltips on regitry pages, etc or should we arbitrarily truncate the summary?
  - For us to extract an implicit summary, we need a syntax that users can rely on, much like rustdoc does.
  - Being consistent with rustdoc would be a benefit.  This most likely means having a markdown parser.  Yes, you can hack it in but that has corner cases, speaking as someone who currently does that in a project (`clap_derive` extracting short-help from a doc-comment)

Reply:
I believe I resolved this in the latest commit between https://github.com/rust-lang/rfcs/blob/97e8301d9f1340be5a05003fb090ad7c7849c971/text/0000-feature-metadata.md?plain=1#L218 and https://github.com/rust-lang/rfcs/blob/97e8301d9f1340be5a05003fb090ad7c7849c971/text/0000-feature-metadata.md?plain=1#L108-L109 (feel free to unresolve if needed)